File: python/selenium/webdriver/support/expected_conditions.py
"""Expected conditions, at the Selenium 4.x import path
``from selenium.webdriver.support import expected_conditions as EC``.

Each function returns a *callable of the driver* suitable for
``WebDriverWait(driver, 10).until(...)``. Signatures and names match mainstream
Selenium so scripts import and call them unchanged::

    from selenium.webdriver.support.wait import WebDriverWait
    from selenium.webdriver.support import expected_conditions as EC
    from selenium.webdriver.common.by import By

    el = WebDriverWait(d, 10).until(
        EC.element_to_be_clickable((By.ID, "submit")))

A *locator* is the mainstream ``(By.X, "value")`` tuple.
"""
from __future__ import annotations

import re
from collections.abc import Iterable
from typing import Callable, Tuple

from ..._webdriver import (
    NoSuchElementException,
    NoSuchFrameException,
    StaleElementReferenceException,
    WebDriverException,
    WebElement,
)

Locator = Tuple[str, str]


def _find(driver, locator: Locator) -> WebElement:
    return driver.find_element(locator[0], locator[1])


def presence_of_element_located(locator: Locator) -> Callable[..., WebElement]:
    """Element present in the DOM (not necessarily visible)."""
    def _cond(driver):
        return _find(driver, locator)
    return _cond


def presence_of_all_elements_located(locator: Locator) -> Callable[..., list]:
    """At least one element present; returns the list."""
    def _cond(driver):
        els = driver.find_elements(locator[0], locator[1])
        return els if els else False
    return _cond


def visibility_of_element_located(locator: Locator) -> Callable[..., WebElement]:
    """Element present AND displayed."""
    def _cond(driver):
        try:
            el = _find(driver, locator)
            return el if el.is_displayed() else False
        except StaleElementReferenceException:
            return False
    return _cond


def visibility_of(element: WebElement) -> Callable[..., WebElement]:
    """A known element becomes displayed."""
    def _cond(_driver):
        try:
            return element if element.is_displayed() else False
        except StaleElementReferenceException:
            return False
    return _cond


def invisibility_of_element_located(locator: Locator) -> Callable[..., bool]:
    """Element is absent or not displayed."""
    def _cond(driver):
        try:
            return not _find(driver, locator).is_displayed()
        except (NoSuchElementException, StaleElementReferenceException):
            return True
    return _cond


def element_to_be_clickable(locator: Locator) -> Callable[..., WebElement]:
    """Element is visible and enabled."""
    def _cond(driver):
        try:
            el = _find(driver, locator)
            return el if (el.is_displayed() and el.is_enabled()) else False
        except StaleElementReferenceException:
            return False
    return _cond


def text_to_be_present_in_element(locator: Locator, text: str) -> Callable[..., bool]:
    """The element's text contains ``text``."""
    def _cond(driver):
        try:
            return text in _find(driver, locator).text
        except StaleElementReferenceException:
            return False
    return _cond


def title_is(title: str) -> Callable[..., bool]:
    return lambda driver: driver.title == title


def title_contains(fragment: str) -> Callable[..., bool]:
    return lambda driver: fragment in driver.title


def url_to_be(url: str) -> Callable[..., bool]:
    return lambda driver: driver.current_url == url


def url_contains(fragment: str) -> Callable[..., bool]:
    return lambda driver: fragment in driver.current_url


def alert_is_present() -> Callable[..., bool]:
    """An alert/confirm/prompt is open."""
    def _cond(driver):
        try:
            driver.alert_text  # raises if no alert
            return True
        except Exception:
            return False
    return _cond


def staleness_of(element: WebElement) -> Callable[..., bool]:
    """Element is no longer attached to the DOM."""
    def _cond(_driver):
        try:
            element.is_enabled()  # any command triggers stale check
            return False
        except StaleElementReferenceException:
            return True
    return _cond


def _element_if_visible(element: WebElement, visibility: bool = True):
    return element if element.is_displayed() == visibility else False


def visibility_of_any_elements_located(locator: Locator) -> Callable[..., list]:
    """At least one matching element is visible; returns the visible ones."""
    def _cond(driver):
        return [e for e in driver.find_elements(*locator) if _element_if_visible(e)]
    return _cond


def visibility_of_all_elements_located(locator: Locator) -> Callable[..., list]:
    """All matching elements are present AND visible; returns the list, else False."""
    def _cond(driver):
        try:
            elements = driver.find_elements(*locator)
            for element in elements:
                if _element_if_visible(element, visibility=False):
                    return False
            return elements
        except StaleElementReferenceException:
            return False
    return _cond


def text_to_be_present_in_element_value(locator: Locator, text_: str) -> Callable[..., bool]:
    """``text_`` is a substring of the element's ``value`` attribute."""
    def _cond(driver):
        try:
            element_text = driver.find_element(*locator).get_attribute("value")
            if element_text is None:
                return False
            return text_ in element_text
        except StaleElementReferenceException:
            return False
    return _cond


def text_to_be_present_in_element_attribute(locator: Locator, attribute_: str, text_: str) -> Callable[..., bool]:
    """``text_`` is a substring of the element's ``attribute_``."""
    def _cond(driver):
        try:
            element_text = driver.find_element(*locator).get_attribute(attribute_)
            if element_text is None:
                return False
            return text_ in element_text
        except StaleElementReferenceException:
            return False
    return _cond


def frame_to_be_available_and_switch_to_it(locator) -> Callable[..., bool]:
    """The frame is available; switch to it. ``locator`` is a (By, value) tuple,
    a name/id string, or a frame WebElement."""
    def _cond(driver):
        try:
            if isinstance(locator, Iterable) and not isinstance(locator, str):
                driver.switch_to.frame(driver.find_element(*locator))
            else:
                driver.switch_to.frame(locator)
            return True
        except NoSuchFrameException:
            return False
    return _cond


def invisibility_of_element(element) -> Callable[..., bool]:
    """Alias of :func:`invisibility_of_element_located` accepting a locator or
    a WebElement."""
    return invisibility_of_element_located(element)


def element_to_be_selected(element: WebElement) -> Callable[..., bool]:
    """The given element is selected."""
    def _cond(_driver):
        return element.is_selected()
    return _cond


def element_located_to_be_selected(locator: Locator) -> Callable[..., bool]:
    """The element found by ``locator`` is selected."""
    def _cond(driver):
        return driver.find_element(*locator).is_selected()
    return _cond


def element_selection_state_to_be(element: WebElement, is_selected: bool) -> Callable[..., bool]:
    """The given element's selection state equals ``is_selected``."""
    def _cond(_driver):
        return element.is_selected() == is_selected
    return _cond


def element_located_selection_state_to_be(locator: Locator, is_selected: bool) -> Callable[..., bool]:
    """The located element's selection state equals ``is_selected``."""
    def _cond(driver):
        try:
            element = driver.find_element(*locator)
            return element.is_selected() == is_selected
        except StaleElementReferenceException:
            return False
    return _cond


def number_of_windows_to_be(num_windows: int) -> Callable[..., bool]:
    """The number of open windows equals ``num_windows``."""
    def _cond(driver):
        return len(driver.window_handles) == num_windows
    return _cond


def new_window_is_opened(current_handles) -> Callable[..., bool]:
    """A new window opened since ``current_handles`` was captured."""
    def _cond(driver):
        return len(driver.window_handles) > len(current_handles)
    return _cond


def element_attribute_to_include(locator: Locator, attribute_: str) -> Callable[..., bool]:
    """The located element has the attribute ``attribute_`` (value not None)."""
    def _cond(driver):
        try:
            element_attribute = driver.find_element(*locator).get_attribute(attribute_)
            return element_attribute is not None
        except StaleElementReferenceException:
            return False
    return _cond


def url_matches(pattern: str) -> Callable[..., bool]:
    """The current URL matches the regular-expression ``pattern`` (re.search)."""
    def _cond(driver):
        return re.search(pattern, driver.current_url) is not None
    return _cond


def url_changes(url: str) -> Callable[..., bool]:
    """The current URL differs from ``url``."""
    def _cond(driver):
        return url != driver.current_url
    return _cond


def any_of(*expected_conditions) -> Callable[..., object]:
    """Logical OR: return the first truthy condition's result, else False."""
    def any_of_condition(driver):
        for expected_condition in expected_conditions:
            try:
                result = expected_condition(driver)
                if result:
                    return result
            except WebDriverException:
                pass
        return False
    return any_of_condition


def all_of(*expected_conditions) -> Callable[..., object]:
    """Logical AND: return the list of each result if all are truthy, else False."""
    def all_of_condition(driver):
        results = []
        for expected_condition in expected_conditions:
            try:
                result = expected_condition(driver)
                if not result:
                    return False
                results.append(result)
            except WebDriverException:
                return False
        return results
    return all_of_condition


def none_of(*expected_conditions) -> Callable[..., bool]:
    """Logical NOR: True when none of the conditions are truthy."""
    def none_of_condition(driver):
        for expected_condition in expected_conditions:
            try:
                result = expected_condition(driver)
                if result:
                    return False
            except WebDriverException:
                pass
        return True
    return none_of_condition


__all__ = [
    "presence_of_element_located",
    "presence_of_all_elements_located",
    "visibility_of_element_located",
    "visibility_of",
    "visibility_of_any_elements_located",
    "visibility_of_all_elements_located",
    "invisibility_of_element_located",
    "invisibility_of_element",
    "element_to_be_clickable",
    "text_to_be_present_in_element",
    "text_to_be_present_in_element_value",
    "text_to_be_present_in_element_attribute",
    "frame_to_be_available_and_switch_to_it",
    "element_to_be_selected",
    "element_located_to_be_selected",
    "element_selection_state_to_be",
    "element_located_selection_state_to_be",
    "number_of_windows_to_be",
    "new_window_is_opened",
    "element_attribute_to_include",
    "title_is",
    "title_contains",
    "url_to_be",
    "url_contains",
    "url_matches",
    "url_changes",
    "any_of",
    "all_of",
    "none_of",
    "alert_is_present",
    "staleness_of",
]

File: java/src/main/java/org/openqa/selenium/support/ui/ExpectedConditions.java
package org.openqa.selenium.support.ui;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;
import org.openqa.selenium.Alert;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.NoAlertPresentException;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.StaleElementReferenceException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

/**
 * Ready-made {@link ExpectedCondition}s for {@link WebDriverWait#until}. Mirrors
 * Selenium 4.x's {@code org.openqa.selenium.support.ui.ExpectedConditions}.
 *
 * <p>A condition returns a truthy value to finish the wait, or {@code null} /
 * {@code false} to keep polling. Conditions that locate an element treat a
 * missing element as "not yet" rather than an error, so they compose.
 */
public final class ExpectedConditions {

    private ExpectedConditions() {}

    /** The element located by {@code locator} is present in the DOM. */
    public static ExpectedCondition<WebElement> presenceOfElementLocated(By locator) {
        return driver -> driver.findElement(locator);
    }

    /** The element located by {@code locator} is present and displayed. */
    public static ExpectedCondition<WebElement> visibilityOfElementLocated(By locator) {
        return driver -> {
            try {
                WebElement el = driver.findElement(locator);
                return el.isDisplayed() ? el : null;
            } catch (NoSuchElementException e) {
                return null;
            }
        };
    }

    /** All elements located by {@code locator} are present (non-empty list). */
    public static ExpectedCondition<List<WebElement>> presenceOfAllElementsLocatedBy(By locator) {
        return driver -> {
            List<WebElement> els = driver.findElements(locator);
            return els.isEmpty() ? null : els;
        };
    }

    /** The element located by {@code locator} is present, displayed and enabled. */
    public static ExpectedCondition<WebElement> elementToBeClickable(By locator) {
        return driver -> {
            try {
                WebElement el = driver.findElement(locator);
                return (el.isDisplayed() && el.isEnabled()) ? el : null;
            } catch (NoSuchElementException e) {
                return null;
            }
        };
    }

    /** The given element is displayed. */
    public static ExpectedCondition<WebElement> visibilityOf(WebElement element) {
        return driver -> element.isDisplayed() ? element : null;
    }

    /** The current page title equals {@code title}. */
    public static ExpectedCondition<Boolean> titleIs(String title) {
        return driver -> title.equals(driver.getTitle());
    }

    /** The current page title contains {@code fraction}. */
    public static ExpectedCondition<Boolean> titleContains(String fraction) {
        return driver -> {
            String t = driver.getTitle();
            return t != null && t.contains(fraction);
        };
    }

    /** The current URL contains {@code fraction}. */
    public static ExpectedCondition<Boolean> urlContains(String fraction) {
        return driver -> {
            String u = driver.getCurrentUrl();
            return u != null && u.contains(fraction);
        };
    }

    /** The element located by {@code locator} contains {@code text}. */
    public static ExpectedCondition<Boolean> textToBePresentInElementLocated(By locator, String text) {
        return driver -> {
            try {
                String elText = driver.findElement(locator).getText();
                return elText != null && elText.contains(text);
            } catch (NoSuchElementException e) {
                return null;
            }
        };
    }

    /** Negate a condition: true when {@code condition} is falsy or raises NoSuchElement. */
    public static ExpectedCondition<Boolean> not(ExpectedCondition<?> condition) {
        return driver -> {
            try {
                Object result = condition.apply(driver);
                return result == null || Boolean.FALSE.equals(result);
            } catch (NoSuchElementException e) {
                return true;
            }
        };
    }

    // ---- titles and URLs ----------------------------------------------------

    /** The current URL equals {@code url}. */
    public static ExpectedCondition<Boolean> urlToBe(String url) {
        return driver -> url.equals(driver.getCurrentUrl());
    }

    /** The current URL matches the regular expression {@code regex}. */
    public static ExpectedCondition<Boolean> urlMatches(String regex) {
        Pattern pattern = Pattern.compile(regex);
        return driver -> {
            String url = driver.getCurrentUrl();
            return url != null && pattern.matcher(url).find();
        };
    }

    // ---- element text -------------------------------------------------------

    /** The given element contains {@code text}. */
    public static ExpectedCondition<Boolean> textToBePresentInElement(WebElement element, String text) {
        return driver -> {
            try {
                String elText = element.getText();
                return elText != null && elText.contains(text);
            } catch (StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** The element located by {@code locator} has exactly the text {@code text}. */
    public static ExpectedCondition<Boolean> textToBe(By locator, String text) {
        return driver -> {
            try {
                return text.equals(driver.findElement(locator).getText());
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** The text of the element located by {@code locator} matches {@code regex}. */
    public static ExpectedCondition<Boolean> textMatches(By locator, Pattern regex) {
        return driver -> {
            try {
                String elText = driver.findElement(locator).getText();
                return elText != null && regex.matcher(elText).find();
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** The {@code value} attribute of the element located by {@code locator} contains {@code text}. */
    public static ExpectedCondition<Boolean> textToBePresentInElementValue(By locator, String text) {
        return driver -> {
            try {
                String value = driver.findElement(locator).getAttribute("value");
                return value != null && value.contains(text);
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** The {@code value} attribute of {@code element} contains {@code text}. */
    public static ExpectedCondition<Boolean> textToBePresentInElementValue(WebElement element, String text) {
        return driver -> {
            try {
                String value = element.getAttribute("value");
                return value != null && value.contains(text);
            } catch (StaleElementReferenceException e) {
                return null;
            }
        };
    }

    // ---- attributes and properties -----------------------------------------

    /** An attribute (or property) of {@code element} equals {@code value}. */
    public static ExpectedCondition<Boolean> attributeToBe(WebElement element, String attribute, String value) {
        return driver -> value.equals(element.getAttribute(attribute));
    }

    /** An attribute (or property) of the located element equals {@code value}. */
    public static ExpectedCondition<Boolean> attributeToBe(By locator, String attribute, String value) {
        return driver -> {
            try {
                return value.equals(driver.findElement(locator).getAttribute(attribute));
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** An attribute (or property) of {@code element} contains {@code value}. */
    public static ExpectedCondition<Boolean> attributeContains(WebElement element, String attribute, String value) {
        return driver -> {
            String actual = element.getAttribute(attribute);
            return actual != null && actual.contains(value);
        };
    }

    /** An attribute (or property) of the located element contains {@code value}. */
    public static ExpectedCondition<Boolean> attributeContains(By locator, String attribute, String value) {
        return driver -> {
            try {
                String actual = driver.findElement(locator).getAttribute(attribute);
                return actual != null && actual.contains(value);
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** An attribute (or property) of {@code element} is set and non-empty. */
    public static ExpectedCondition<Boolean> attributeToBeNotEmpty(WebElement element, String attribute) {
        return driver -> {
            String actual = element.getAttribute(attribute);
            return actual != null && !actual.isEmpty();
        };
    }

    /** The literal DOM attribute of {@code element} equals {@code value}. */
    public static ExpectedCondition<Boolean> domAttributeToBe(WebElement element, String attribute, String value) {
        return driver -> value.equals(element.getDomAttribute(attribute));
    }

    /** The DOM property of {@code element} equals {@code value}. */
    public static ExpectedCondition<Boolean> domPropertyToBe(WebElement element, String property, String value) {
        return driver -> value.equals(element.getDomProperty(property));
    }

    // ---- visibility and invisibility ---------------------------------------

    /** Every element in {@code elements} is displayed; yields the list. */
    public static ExpectedCondition<List<WebElement>> visibilityOfAllElements(List<WebElement> elements) {
        return driver -> {
            for (WebElement el : elements) {
                if (!el.isDisplayed()) {
                    return null;
                }
            }
            return elements.isEmpty() ? null : elements;
        };
    }

    /** Every element in {@code elements} is displayed; yields the list. */
    public static ExpectedCondition<List<WebElement>> visibilityOfAllElements(WebElement... elements) {
        return visibilityOfAllElements(Arrays.asList(elements));
    }

    /** All elements located by {@code locator} are present and displayed. */
    public static ExpectedCondition<List<WebElement>> visibilityOfAllElementsLocatedBy(By locator) {
        return driver -> {
            List<WebElement> els = driver.findElements(locator);
            for (WebElement el : els) {
                if (!el.isDisplayed()) {
                    return null;
                }
            }
            return els.isEmpty() ? null : els;
        };
    }

    /** The elements found by {@code childLocator} under {@code parent} are displayed. */
    public static ExpectedCondition<List<WebElement>> visibilityOfNestedElementsLocatedBy(
            By parent, By childLocator) {
        return driver -> {
            List<WebElement> els = driver.findElement(parent).findElements(childLocator);
            for (WebElement el : els) {
                if (!el.isDisplayed()) {
                    return null;
                }
            }
            return els.isEmpty() ? null : els;
        };
    }

    /** The given element is not displayed (or is gone from the DOM). */
    public static ExpectedCondition<Boolean> invisibilityOf(WebElement element) {
        return driver -> {
            try {
                return !element.isDisplayed();
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return true;
            }
        };
    }

    /** No element located by {@code locator} is displayed (or none is present). */
    public static ExpectedCondition<Boolean> invisibilityOfElementLocated(By locator) {
        return driver -> {
            try {
                return !driver.findElement(locator).isDisplayed();
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return true;
            }
        };
    }

    /** The located element is gone, hidden, or no longer carries {@code text}. */
    public static ExpectedCondition<Boolean> invisibilityOfElementWithText(By locator, String text) {
        return driver -> {
            try {
                return !text.equals(driver.findElement(locator).getText());
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return true;
            }
        };
    }

    /** None of {@code elements} is displayed. */
    public static ExpectedCondition<Boolean> invisibilityOfAllElements(List<WebElement> elements) {
        return driver -> {
            for (WebElement el : elements) {
                try {
                    if (el.isDisplayed()) {
                        return false;
                    }
                } catch (NoSuchElementException | StaleElementReferenceException e) {
                    // gone counts as invisible
                }
            }
            return true;
        };
    }

    /** None of {@code elements} is displayed. */
    public static ExpectedCondition<Boolean> invisibilityOfAllElements(WebElement... elements) {
        return invisibilityOfAllElements(Arrays.asList(elements));
    }

    // ---- counting -----------------------------------------------------------

    /** Exactly {@code number} elements match {@code locator}. */
    public static ExpectedCondition<List<WebElement>> numberOfElementsToBe(By locator, Integer number) {
        return driver -> {
            List<WebElement> els = driver.findElements(locator);
            return els.size() == number ? els : null;
        };
    }

    /** More than {@code number} elements match {@code locator}. */
    public static ExpectedCondition<List<WebElement>> numberOfElementsToBeMoreThan(By locator, Integer number) {
        return driver -> {
            List<WebElement> els = driver.findElements(locator);
            return els.size() > number ? els : null;
        };
    }

    /** Fewer than {@code number} elements match {@code locator}. */
    public static ExpectedCondition<List<WebElement>> numberOfElementsToBeLessThan(By locator, Integer number) {
        return driver -> {
            List<WebElement> els = driver.findElements(locator);
            return els.size() < number ? els : null;
        };
    }

    /** The session has exactly {@code expected} top-level windows/tabs. */
    public static ExpectedCondition<Boolean> numberOfWindowsToBe(int expected) {
        return driver -> driver.getWindowHandles().size() == expected;
    }

    // ---- nested lookups -----------------------------------------------------

    /** An element matching {@code child} exists under the element matching {@code parent}. */
    public static ExpectedCondition<WebElement> presenceOfNestedElementLocatedBy(By parent, By child) {
        return driver -> {
            try {
                return driver.findElement(parent).findElement(child);
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** Elements matching {@code child} exist under the element matching {@code parent}. */
    public static ExpectedCondition<List<WebElement>> presenceOfNestedElementsLocatedBy(By parent, By child) {
        return driver -> {
            try {
                List<WebElement> els = driver.findElement(parent).findElements(child);
                return els.isEmpty() ? null : els;
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    // ---- selection state ----------------------------------------------------

    /** The given element is selected. */
    public static ExpectedCondition<Boolean> elementToBeSelected(WebElement element) {
        return elementSelectionStateToBe(element, true);
    }

    /** The given element's selected state is {@code selected}. */
    public static ExpectedCondition<Boolean> elementSelectionStateToBe(WebElement element, boolean selected) {
        return driver -> element.isSelected() == selected;
    }

    /** The located element's selected state is {@code selected}. */
    public static ExpectedCondition<Boolean> elementSelectionStateToBe(By locator, boolean selected) {
        return driver -> {
            try {
                return driver.findElement(locator).isSelected() == selected;
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** The element located by {@code locator} is present, displayed and enabled. */
    public static ExpectedCondition<WebElement> elementToBeClickable(WebElement element) {
        return driver -> (element.isDisplayed() && element.isEnabled()) ? element : null;
    }

    // ---- staleness, frames, alerts -----------------------------------------

    /** The given element is no longer attached to the DOM. */
    public static ExpectedCondition<Boolean> stalenessOf(WebElement element) {
        return driver -> {
            try {
                element.isEnabled();
                return false;
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return true;
            }
        };
    }

    /** Re-run {@code condition}, retrying once if the element under it went stale. */
    public static <T> ExpectedCondition<T> refreshed(ExpectedCondition<T> condition) {
        return driver -> {
            try {
                return condition.apply(driver);
            } catch (StaleElementReferenceException e) {
                return condition.apply(driver);
            }
        };
    }

    /** The frame is available; switches to it and yields the driver. */
    public static ExpectedCondition<WebDriver> frameToBeAvailableAndSwitchToIt(By locator) {
        return driver -> {
            try {
                return driver.switchTo().frame(driver.findElement(locator));
            } catch (NoSuchElementException | StaleElementReferenceException e) {
                return null;
            }
        };
    }

    /** The frame with this name or id is available; switches to it. */
    public static ExpectedCondition<WebDriver> frameToBeAvailableAndSwitchToIt(String nameOrId) {
        return driver -> {
            try {
                return driver.switchTo().frame(nameOrId);
            } catch (RuntimeException e) {
                return null;
            }
        };
    }

    /** The frame at this index is available; switches to it. */
    public static ExpectedCondition<WebDriver> frameToBeAvailableAndSwitchToIt(int index) {
        return driver -> {
            try {
                return driver.switchTo().frame(index);
            } catch (RuntimeException e) {
                return null;
            }
        };
    }

    /** A JavaScript alert is showing; yields it. */
    public static ExpectedCondition<Alert> alertIsPresent() {
        return driver -> {
            try {
                return driver.switchTo().alert();
            } catch (NoAlertPresentException e) {
                return null;
            }
        };
    }

    // ---- script -------------------------------------------------------------

    /** {@code script} runs without throwing. */
    public static ExpectedCondition<Boolean> javaScriptThrowsNoExceptions(String script) {
        return driver -> {
            try {
                ((JavascriptExecutor) driver).executeScript(script);
                return true;
            } catch (RuntimeException e) {
                return false;
            }
        };
    }

    /** {@code script} returns a non-null, non-empty value; yields it. */
    public static ExpectedCondition<Object> jsReturnsValue(String script) {
        return driver -> {
            Object value = ((JavascriptExecutor) driver).executeScript(script);
            if (value instanceof List) {
                return ((List<?>) value).isEmpty() ? null : value;
            }
            if (value instanceof String) {
                return ((String) value).isEmpty() ? null : value;
            }
            return value;
        };
    }

    // ---- composition --------------------------------------------------------

    /** Every condition holds. */
    public static ExpectedCondition<Boolean> and(ExpectedCondition<?>... conditions) {
        return driver -> {
            for (ExpectedCondition<?> condition : conditions) {
                Object result = condition.apply(driver);
                if (result == null || Boolean.FALSE.equals(result)) {
                    return false;
                }
            }
            return true;
        };
    }

    /** At least one condition holds. */
    public static ExpectedCondition<Boolean> or(ExpectedCondition<?>... conditions) {
        return driver -> {
            for (ExpectedCondition<?> condition : conditions) {
                try {
                    Object result = condition.apply(driver);
                    if (result != null && !Boolean.FALSE.equals(result)) {
                        return true;
                    }
                } catch (RuntimeException e) {
                    // a throwing condition simply does not hold
                }
            }
            return false;
        };
    }
}
